javascript/javascript3/week2/visual/main.js

```diff
@@ -0,0 +1,35 @@
+let red = document.getElementById("red");
+let blue = document.getElementById("blue");
+let green = document.getElementById("green");
+
+function translateOneByOne() {
+    moveElement(red, { x: 20, y: 300 })
+    .then(() => {
+    console.log('Red circle moved');
+    moveElement(blue, { x: 400, y: 300 })
+
+    .then(() => {
+    console.log('Blue circle moved');
+    moveElement(green, { x: 400, y: 20 })
+    
+    .then(() => {
+    console.log('Green circle moved');
+    });
+    });
+    });
+
+}

```

Comment:
This is an example of 'callback hell', it would look similar if you had used callback functions instead of promises. It works fine though, but this is the problem the that promises can solve compared to callbacks.

Ìf i parse the syntex (indention) correctly, it looks like this:
```
function translateOneByOne() {
  moveElement(red, { x: 20, y: 300 })
  .then(() => {
    console.log('Red circle moved');

    moveElement(blue, { x: 400, y: 300 })
    .then(() => {
      console.log('Blue circle moved');
      
      moveElement(green, { x: 400, y: 20 })
      .then(() => {
        console.log('Green circle moved');
      });
    });
  });
}
```

When using promises, you can `return` the promise to the next .then instead. So this could (should) be re-written as:

```
function translateOneByOne() {
  moveElement(red, { x: 20, y: 300 })
  .then(() => {
    console.log('Red circle moved');
    return moveElement(blue, { x: 400, y: 300 })
  .then(() => {
    console.log('Blue circle moved');
    return moveElement(green, { x: 400, y: 20 })
  .then(() => {
    console.log('Green circle moved');
  });
}
```

See how i return the promise each time, and it get picked up in the next `.then` - called chaining. It makes it more readable/maintainable.